File: src/game/scripts/vscripts/abilities/items/sange_kaya_yasha.lua

-- Lua Library Imports
LinkLuaModifier("modifier_item_sange_kaya_yasha","abilities/items/sange_kaya_yasha.lua",LUA_MODIFIER_MOTION_NONE);
LinkLuaModifier("modifier_item_sange_kaya_yasha_buffs","abilities/items/sange_kaya_yasha.lua",LUA_MODIFIER_MOTION_NONE);
item_sange_kaya_yasha_consumable = item_sange_kaya_yasha_consumable or item_consumable_redux.new()
item_sange_kaya_yasha_consumable.__index = item_sange_kaya_yasha_consumable
item_sange_kaya_yasha_consumable.__base = item_consumable_redux
function item_sange_kaya_yasha_consumable.new(construct, ...)
    local self = setmetatable({}, item_sange_kaya_yasha_consumable)
    if construct and item_sange_kaya_yasha_consumable.constructor then item_sange_kaya_yasha_consumable.constructor(self, ...) end
    return self
end
function item_sange_kaya_yasha_consumable.constructor(self)
end
function item_sange_kaya_yasha_consumable.GetIntrinsicModifierName(self)
    return "modifier_item_sange_kaya_yasha"
end
modifier_item_sange_kaya_yasha_consumable = modifier_item_sange_kaya_yasha_consumable or {}
modifier_item_sange_kaya_yasha_consumable.__index = modifier_item_sange_kaya_yasha_consumable
function modifier_item_sange_kaya_yasha_consumable.new(construct, ...)
    local self = setmetatable({}, modifier_item_sange_kaya_yasha_consumable)
    if construct and modifier_item_sange_kaya_yasha_consumable.constructor then modifier_item_sange_kaya_yasha_consumable.constructor(self, ...) end
    return self
end
function modifier_item_sange_kaya_yasha_consumable.constructor(self)
end
function modifier_item_sange_kaya_yasha_consumable.IsHidden(self)
    if (not self:GetAbility()) then
        self:Destroy();
        return false
    end
    return self:GetAbility().IsItem==nil
end
function modifier_item_sange_kaya_yasha_consumable.IsPermanent(self)
    return true
end
function modifier_item_sange_kaya_yasha_consumable.GetTexture(self)
    return "item_sange_kaya_yasha"
end
function modifier_item_sange_kaya_yasha_consumable.GetAttributes(self)
    return MODIFIER_ATTRIBUTE_MULTIPLE
end
function modifier_item_sange_kaya_yasha_consumable.OnCreated(self)
    if IsServer() then
        self:GetParent():AddNewModifier(self:GetParent(),self:GetAbility(),"modifier_item_sange_kaya_yasha_buffs",{});
    end
end
function modifier_item_sange_kaya_yasha_consumable.DeclareFunctions(self)
    return {MODIFIER_PROPERTY_STATS_INTELLECT_BONUS,MODIFIER_PROPERTY_STATS_AGILITY_BONUS,MODIFIER_PROPERTY_STATS_STRENGTH_BONUS,MODIFIER_PROPERTY_PREATTACK_BONUS_DAMAGE,MODIFIER_PROPERTY_ATTACKSPEED_BONUS_CONSTANT}
end
function modifier_item_sange_kaya_yasha_consumable.GetModifierBonusStats_Intellect(self)
    if (not self:GetAbility()) then
        self:Destroy();
        return 0
    end
    return self:GetAbility():GetSpecialValueFor("sange_kaya_yasha_bonus_intellect")
end
function modifier_item_sange_kaya_yasha_consumable.GetModifierBonusStats_Agility(self)
    if (not self:GetAbility()) then
        self:Destroy();
        return 0
    end
    return self:GetAbility():GetSpecialValueFor("sange_kaya_yasha_bonus_agility")
end
function modifier_item_sange_kaya_yasha_consumable.GetModifierBonusStats_Strength(self)
    if (not self:GetAbility()) then
        self:Destroy();
        return 0
    end
    return self:GetAbility():GetSpecialValueFor("sange_kaya_yasha_bonus_strength")
end
function modifier_item_sange_kaya_yasha_consumable.GetModifierPreAttack_BonusDamage(self)
    if (not self:GetAbility()) then
        self:Destroy();
        return 0
    end
    return self:GetAbility():GetSpecialValueFor("sange_kaya_yasha_bonus_damage")
end
function modifier_item_sange_kaya_yasha_consumable.GetModifierAttackSpeedBonus_Constant(self)
    if (not self:GetAbility()) then
        self:Destroy();
        return 0
    end
    return self:GetAbility():GetSpecialValueFor("sange_kaya_yasha_bonus_attack_speed")
end
modifier_item_sange_kaya_yasha_consumable_buffs = modifier_item_sange_kaya_yasha_consumable_buffs or {}
modifier_item_sange_kaya_yasha_consumable_buffs.__index = modifier_item_sange_kaya_yasha_consumable_buffs
function modifier_item_sange_kaya_yasha_consumable_buffs.new(construct, ...)
    local self = setmetatable({}, modifier_item_sange_kaya_yasha_consumable_buffs)
    if construct and modifier_item_sange_kaya_yasha_consumable_buffs.constructor then modifier_item_sange_kaya_yasha_consumable_buffs.constructor(self, ...) end
    return self
end
function modifier_item_sange_kaya_yasha_consumable_buffs.constructor(self)
end
function modifier_item_sange_kaya_yasha_consumable_buffs.IsPermanent(self)
    return true
end
function modifier_item_sange_kaya_yasha_consumable_buffs.IsHidden(self)
    return true
end
function modifier_item_sange_kaya_yasha_consumable_buffs.DeclareFunctions(self)
    return {MODIFIER_PROPERTY_STATUS_RESISTANCE_STACKING,MODIFIER_PROPERTY_MOVESPEED_BONUS_PERCENTAGE,MODIFIER_PROPERTY_SPELL_AMPLIFY_PERCENTAGE,MODIFIER_PROPERTY_MANACOST_PERCENTAGE_STACKING}
end
function modifier_item_sange_kaya_yasha_consumable_buffs.GetModifierStatusResistanceStacking(self)
    if (not self:GetAbility()) then
        self:Destroy();
        return 0
    end
    return self:GetAbility():GetSpecialValueFor("sange_kaya_yasha_status_resistance")
end
function modifier_item_sange_kaya_yasha_consumable_buffs.GetModifierMoveSpeedBonus_Percentage(self)
    if (not self:GetAbility()) then
        self:Destroy();
        return 0
    end
    return self:GetAbility():GetSpecialValueFor("sange_kaya_yasha_movement_speed_percent_bonus")
end
function modifier_item_sange_kaya_yasha_consumable_buffs.GetModifierSpellAmplify_Percentage(self)
    if (not self:GetAbility()) then
        self:Destroy();
        return 0
    end
    return self:GetAbility():GetSpecialValueFor("sange_kaya_yasha_spell_amp")
end
function modifier_item_sange_kaya_yasha_consumable_buffs.GetModifierPercentageManacostStacking(self)
    if (not self:GetAbility()) then
        self:Destroy();
        return 0
    end
    return self:GetAbility():GetSpecialValueFor("sange_kaya_yasha_manacost_reduction")
end

File: src/game/scripts/vscripts/abilities/items/sange_kaya_yasha.ts
//require('abilities/items/consumable_baseclass')
LinkLuaModifier("modifier_item_sange_kaya_yasha","abilities/items/sange_kaya_yasha.lua",LuaModifierType.LUA_MODIFIER_MOTION_NONE)
LinkLuaModifier("modifier_item_sange_kaya_yasha_buffs","abilities/items/sange_kaya_yasha.lua",LuaModifierType.LUA_MODIFIER_MOTION_NONE)

class item_sange_kaya_yasha_consumable extends item_consumable_redux {
	GetIntrinsicModifierName() { 
		return "modifier_item_sange_kaya_yasha";
	}
}

class modifier_item_sange_kaya_yasha_consumable extends CDOTA_Modifier_Lua {
	IsHidden() {
		if (!this.GetAbility()) {
			this.Destroy();
			return false;
		}
		return this.GetAbility().IsItem == null;
	}

	IsPermanent() {return true;}
	GetTexture() {return "item_sange_kaya_yasha";}

	GetAttributes() { 
		return DOTAModifierAttribute_t.MODIFIER_ATTRIBUTE_MULTIPLE
	}

	OnCreated() {
		if (IsServer()) {
			this.GetParent().AddNewModifier(this.GetParent(),this.GetAbility(),"modifier_item_sange_kaya_yasha_buffs",{})
		}
	}

	DeclareFunctions() {
		return [
			modifierfunction.MODIFIER_PROPERTY_STATS_INTELLECT_BONUS,
			modifierfunction.MODIFIER_PROPERTY_STATS_AGILITY_BONUS,
			modifierfunction.MODIFIER_PROPERTY_STATS_STRENGTH_BONUS,
			modifierfunction.MODIFIER_PROPERTY_PREATTACK_BONUS_DAMAGE,
			modifierfunction.MODIFIER_PROPERTY_ATTACKSPEED_BONUS_CONSTANT,
		]
	}

	GetModifierBonusStats_Intellect() {
		if (!this.GetAbility()) {
			this.Destroy();
			return 0;
		}
		return this.GetAbility().GetSpecialValueFor("sange_kaya_yasha_bonus_intellect")
	}

	GetModifierBonusStats_Agility() {
		if (!this.GetAbility()) {
			this.Destroy();
			return 0;
		}
		return this.GetAbility().GetSpecialValueFor("sange_kaya_yasha_bonus_agility")
	}

	GetModifierBonusStats_Strength() {
		if (!this.GetAbility()) {
			this.Destroy();
			return 0;
		}
		return this.GetAbility().GetSpecialValueFor("sange_kaya_yasha_bonus_strength")
	}

	GetModifierPreAttack_BonusDamage() {
		if (!this.GetAbility()) {
			this.Destroy();
			return 0;
		}
		return this.GetAbility().GetSpecialValueFor("sange_kaya_yasha_bonus_damage")
	}

	GetModifierAttackSpeedBonus_Constant() {
		if (!this.GetAbility()) {
			this.Destroy();
			return 0;
		}
		return this.GetAbility().GetSpecialValueFor("sange_kaya_yasha_bonus_attack_speed")
	}
}


class modifier_item_sange_kaya_yasha_consumable_buffs extends CDOTA_Modifier_Lua {
	IsPermanent() {return true;}
	IsHidden() {return true;}

	DeclareFunctions() {
		return [
			modifierfunction.MODIFIER_PROPERTY_STATUS_RESISTANCE_STACKING,
			modifierfunction.MODIFIER_PROPERTY_MOVESPEED_BONUS_PERCENTAGE,
			modifierfunction.MODIFIER_PROPERTY_SPELL_AMPLIFY_PERCENTAGE,
			modifierfunction.MODIFIER_PROPERTY_MANACOST_PERCENTAGE_STACKING,
		]
	}

	GetModifierStatusResistanceStacking() {
		if (!this.GetAbility()) {
			this.Destroy();
			return 0;
		}
		return this.GetAbility().GetSpecialValueFor("sange_kaya_yasha_status_resistance")
	}

	GetModifierMoveSpeedBonus_Percentage() {
		if (!this.GetAbility()) {
			this.Destroy();
			return 0;
		}
		return this.GetAbility().GetSpecialValueFor("sange_kaya_yasha_movement_speed_percent_bonus")
	}

	GetModifierSpellAmplify_Percentage() {
		if (!this.GetAbility()) {
			this.Destroy();
			return 0;
		}
		return this.GetAbility().GetSpecialValueFor("sange_kaya_yasha_spell_amp")
	}

	GetModifierPercentageManacostStacking() {
		if (!this.GetAbility()) {
			this.Destroy();
			return 0;
		}
		return this.GetAbility().GetSpecialValueFor("sange_kaya_yasha_manacost_reduction")
	}

	
}

// "item_recipe_sange_kaya_yasha_consumable"
//     {
//         // General
//         //-------------------------------------------------------------------------------------------------------------
//         "ID"                            "99997"                                                       // unique ID number for this item.  Do not change this once established or it will invalidate collected stats.
//         "Model"                         "models/props_gameplay/recipe.vmdl"
        
//         // Item Info
//         //-------------------------------------------------------------------------------------------------------------
//         "ItemCost"                      "0"   
//         "ItemShopTags"                  ""
        
//         // Recipe
//         //-------------------------------------------------------------------------------------------------------------
//         "ItemRecipe"                    "1"
//         "ItemResult"                    "item_sange_kaya_yasha_consumable"
//         "ItemRequirements"
//         {
            
//             "01"                        "item_sange_and_yasha;item_kaya"
//             "02"                        "item_kaya_and_sange;item_yasha"
//             "03"                        "item_yasha_and_kaya;item_sange"
//             "04"                        "item_sange;item_kaya;item_yasha"
//         }  
//     }
//     "item_sange_kaya_yasha_consumable"
//     {
//         "ID"                            "99998"
//         "ScriptFile"                    "abilities/items/sange_kaya_yasha.lua"
//         "AbilityTextureName"            "item_bloodstone"
//         "BaseClass"                     "item_lua"
//         "ItemCost"                      "6150"

//         "AbilityBehavior"               "DOTA_ABILITY_BEHAVIOR_UNIT_TARGET | DOTA_ABILITY_BEHAVIOR_IMMEDIATE"
//         "AbilityUnitTargetTeam"         "DOTA_UNIT_TARGET_TEAM_FRIENDLY"
//         "AbilityUnitTargetType"         "DOTA_UNIT_TARGET_HERO"
//         "AbilityGoldCost"               "6150"
//         "ItemPurchasable"              "0 "

         

